aggregation/src/lib.rs

```diff
@@ -52,45 +53,309 @@ alloy_sol_types::sol! {
+    /// [MerkleMountainRange] is empty.
+    pub fn finalized_root(mut self) -> Option<Digest> {
+        match self.is_empty() {
+            true => None,
+            false => {
+                // finalize should only fail if the MMR is empty.
+                self.finalize().unwrap();
+                Some(self.0[0].digest)
+            }
+        }
+    }
+
+    /// Returns true if the [MerkleMountainRange] is finalized. No new nodes can be added to a
+    /// finalized [MerkleMountainRange].
+    pub fn is_finalized(&self) -> bool {
+        // If the peak has a max-depth of 255, then the mountain range is finalized and no new
+        // peaks can be pushed to it. Note that this state can only be achieved by calling
+        // `finalize` since it is computationally infeasible to push 2^256 nodes onto the MMR,
+        // although it is theoretically consistent that if an MMR reached a state of being a
+        // single peak with max depth value of 255, it would be naturally finalized.
+        self.0
+            .first()
+            .map(|peak| peak.max_depth == u8::MAX)
+            .unwrap_or(false)
+    }
+
+    /// Returns true if the [MerkleMountainRange] is empty.
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// A compact encoding for the [MerkleMountainRange] used in encoding the journal designed to
+    /// be efficient for use in the EVM, and designed to ensure it is possible to construct the
+    /// journal encoding of a finalized [MerkleMountainRange] given only the finalized root.
+    pub fn encode(&self) -> Vec<u8> {
+        // bitmap encodes the max-depth values present in the MerkleMountainRange. Note that when
+        // finalized, the bitmap is guaranteed to be equal to 1 << 255.
+        let mut bitmap = U256::ZERO;
+        let mut peaks = Vec::<Digest>::with_capacity(self.0.len());
+        // Iterate over the peaks from greatest to least max-depth.
+        for peak in self.0.iter() {
+            bitmap.set_bit(peak.max_depth as usize, true);
+            peaks.push(peak.digest);
         }
+        [&bitmap.as_le_bytes(), bytemuck::cast_slice(&peaks)].concat()
     }
 
-    #[inline]
-    pub fn root(&self) -> Digest {
-        match self {
-            GuestInput::Singleton { claim, .. } => claim.digest::<sha::Impl>(),
-            GuestInput::Join {
-                left_set_root,
-                right_set_root,
-                ..
-            } => commutative_keccak256(left_set_root, right_set_root),
+    /// Decode the specialized journal encoding. See [MerkleMountainRange::encode].
+    pub fn decode(bytes: impl AsRef<[u8]>) -> Result<Self, DecodingError> {
+        // Read the first 32 bytes as the bitmap.
+        let (mut chunk, mut bytes) = bytes
+            .as_ref()
+            .split_at_checked(U256::BYTES)
+            .ok_or(DecodingError::UnexpectedEnd)?;
+        let bitmap = U256::from_le_slice(chunk);
+        if bitmap > (U256::from(1) << u8::MAX) {
```

Comment:
Does the Compiler understands that this is a constant or will it be runtime evaluated?
I would go with `uint!(1_U256 << u8::MAX)` for added clarity.

Reply:
Nice! I didn't know about that macro.
6627c08